InvoiceNumbering/Domain.fs

```diff
@@ -0,0 +1,62 @@
+﻿module InvoiceNumbering
+
+module Events =
+  type Event =
+    | InvoiceNumberReserved of
+      {| invoiceNumber: int
+         reservedFor: Invoice.InvoiceId |}
+    | Snapshotted of
+      {| next: int
+         allocated: Map<Invoice.InvoiceId, int> |}
+
+    interface TypeShape.UnionContract.IUnionContract
+
+  let codec = FsCodec.SystemTextJson.Codec.Create<Event>()
+
+module Fold =
+  open Events
+
+  type State =
+    { next: int
+      allocated: Map<Invoice.InvoiceId, int> }
+
+  let initial: State = { next = 1; allocated = Map.empty }
+
+  let evolve state event =
+    match event with
+    | Snapshotted data ->
+      { next = data.next
+        allocated = data.allocated }
+    | InvoiceNumberReserved data ->
+      { next = data.invoiceNumber + 1
+        allocated = Map.add data.reservedFor data.invoiceNumber state.allocated }
+
+  let fold: State -> Event seq -> State = Seq.fold evolve
+  let toSnapshot s = Snapshotted {| next = s.next; allocated = s.allocated |}
+  let snapshotEventType = nameof Snapshotted
+
+module Decide =
+  open Fold
+
+  let reserve reservedFor state =
+    match Map.tryFind reservedFor state.allocated with
+    | None ->
+      state.next,
+      [ Events.InvoiceNumberReserved
+          {| invoiceNumber = state.next
+             reservedFor = reservedFor |} ]
+    | Some n -> n, []
+
+[<Literal>]
```

Comment:
convention says top of file
- provides a cue that a) this is an event sourced aggregate
- makes it easier to scan/navigate

Reply:
on second thought, I think I'll keep it this way since this project is pedagogical. The Invoice domain file also has it in this order, and that's purely because when I wrote the post the category/stream don't factor into the discussion until later.

Reply:
https://github.com/jet/dotnet-templates#aggregate-module
Yes definitely debatable in context of this sample
If you start putting a `module Config` per aggregate, then it also can get noisy
A final reason to have it up top by convention is in wierd cases where you have multiple folds and/or services (read/write/index etc) they all share a category name and its nice for that to have an ultra stable/obvious place e.g. https://github.com/jet/propulsion/blob/master/src/Propulsion.DynamoStore/AppendsIndex.fs#L61